File: src/main/java/com/buffer/lorena/bot/handler/MessageHandler.java
package com.buffer.lorena.bot.handler;

import com.buffer.lorena.bot.service.LorenaService;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.javacord.api.entity.server.Server;
import org.javacord.api.event.message.MessageCreateEvent;
import org.javacord.api.listener.message.MessageCreateListener;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Locale;

/**
 * The type Message handler.
 */
@Component
public class MessageHandler implements MessageCreateListener {
    private final Logger logger = LogManager.getLogger(MessageHandler.class);
    private static final String LORENA_TEXT = "lorena";

    private LorenaService lorenaService;
    private SlashCommandHandler slashCommandHandler;

    /**
     * Instantiates a new Message handler.
     */
    public MessageHandler() {
    }

    /**
     * Instantiates a new Message handler.
     *
     * @param lorenaService       the lorena service
     * @param slashCommandHandler the slash command handler
     */
    @Autowired
    public MessageHandler(LorenaService lorenaService, SlashCommandHandler slashCommandHandler) {
        this.lorenaService = lorenaService;
        this.slashCommandHandler = slashCommandHandler;
    }

    /**
     * On message create.
     *
     * @param event the event
     */
    @Override
    public void onMessageCreate(MessageCreateEvent event) {
        logger.info("message received in server {}: {}", event.getServer().map(Server::getName).get(), event.getMessageContent());
        slashCommandHandler.registerSlashCommands(event.getServer().get());

        String[] parsedMessage = event.getMessageContent().split(" ");
        if ((parsedMessage[0].equalsIgnoreCase("!lore") || parsedMessage[0].equalsIgnoreCase("!lorebot")) && !event.getMessageAuthor().isBotUser()) {
            switch (parsedMessage[1].toLowerCase(Locale.ROOT)) {
                case "ping":
                    event.getChannel().sendMessage("Pong!");
                    break;
                case "votes":
                    this.lorenaService.changeUserVoteThreshold(event, parsedMessage[2]);
                    break;
                case "setlorechannel":
                    this.lorenaService.setServerLoreChannel(event, parsedMessage[2]);
                    break;
                case "setsuggestionchannel":
                    this.lorenaService.setServerSuggestionChannel(event, parsedMessage[2]);
                    break;
                case "dolore":
                    this.lorenaService.sendRandomLore(event);
                    break;
                default:
                    event.getChannel().sendMessage("fuck off");
                    break;
            }
        } else if(event.getMessageContent().toLowerCase(Locale.ROOT).contains(LORENA_TEXT)){
            this.lorenaService.sendRandomLore(event);
        }
    }
}

File: src/main/kotlin/com/buffer/lorena/handler/MessageHandler.kt
package com.buffer.lorena.handler

import com.buffer.lorena.bot.converter.LorenaConverter
import org.javacord.api.listener.message.MessageCreateListener
import com.buffer.lorena.bot.service.LorenaService
import com.buffer.lorena.service.RedditService
import com.buffer.lorena.service.UnitConversionService
import org.apache.logging.log4j.LogManager
import org.javacord.api.entity.server.Server
import org.javacord.api.event.message.MessageCreateEvent
import org.springframework.stereotype.Component
import java.util.Locale

/**
 * The type Message handler.
 */
@Component
class MessageHandler(
        private val lorenaService: LorenaService,
        private val redditService: RedditService,
        private val unitConversionService: UnitConversionService,
        private val lorenaConverter: LorenaConverter
) : MessageCreateListener {
    private val logger = LogManager.getLogger(MessageHandler::class.java)

    /**
     * On message create.
     *
     * @param event the event
     */
    override fun onMessageCreate(event: MessageCreateEvent) {
        logger.info("New message in {}(#{}) by {}({}): {}",
            event.server.map { obj: Server -> obj.name }.get(),
            event.channel.asServerTextChannel().get().name,
            event.messageAuthor.name,
            event.messageAuthor.displayName,
            event.messageContent + if (event.message.attachments.isNotEmpty()) " " + event.message.attachments.map { obj -> obj.url.toString() } else "")

        val parsedMessage = event.messageContent.split(" ")

        lorenaConverter.convertServer(event.server.get())

        when {
            (parsedMessage[0].equals("!lore", ignoreCase = true)
                    || parsedMessage[0].equals("!lorebot", ignoreCase = true))
                    && !event.messageAuthor.isBotUser ->
                when (parsedMessage[1].lowercase(Locale.ROOT)) {
                    "ping" -> event.channel.sendMessage("Pong!")
                    "votes" -> lorenaService.changeUserVoteThreshold(event, parsedMessage[2])
                    "setlorechannel" -> lorenaService.setServerLoreChannel(event, parsedMessage[2])
                    "setsuggestionchannel" -> lorenaService.setServerSuggestionChannel(event, parsedMessage[2])
                    "dolore" -> lorenaService.sendRandomLore(event)
                    else -> event.channel.sendMessage("fuck off")
                }

            (event.messageContent.lowercase().contains(LORENA_TEXT) || event.messageContent.lowercase().contains(LORENZO_TEXT)) -> {
                logger.info("detected a new summon")
                if(event.messageContent.lowercase().contains("nuggies")
                    && event.messageAuthor.name.contains("tina")) {
                    event.channel.sendMessage("nuggies machine broke")
                }
//                else {
//                    repeat(Random().nextInt(2)+1) {
//                        lorenaService.sendRandomLore(event)
//                    }
//                }
            }
        }

        parsedMessage.filter { it.isReddit() }.forEach {
            redditService.sendLink(it, event)
        }

        unitConversionService.parseMessage(event.messageContent, event)
    }

    private fun String.isReddit(): Boolean = this.matches(REDDIT_REGEX)

    companion object {
        private const val LORENA_TEXT = "lorena"
        private const val LORENZO_TEXT = "lorenzo"
        private val REDDIT_REGEX = """\b([ru])/([a-zA-Z0-9-_]{3,21})\b""".toRegex()
    }
}
